### ISSUE TITLE
cabal tries to build package with incompatible dependencies

### ISSUE THREAD

**Issue by chshersh** [haskell/cabal#6166]
**Describe the bug**

Sometimes bulds on CI fail for a very strange reason even locally everything passes. Cabal tries to find plan and uses incompatible versions of libraries. Here is the error when trying to build `htoml-megaparsec` dependency for `tomland` library:

```
Failed to build htoml-megaparsec-2.1.0.3.
Build log (
/home/travis/.cabal/logs/ghc-8.6.5/htoml-megaparsec-2.1.0.3-4b4579106a015b896d103c21db8f2fac3f5d24b1ecbd2f4464c4ae7e70f1d197.log
):
Configuring library for htoml-megaparsec-2.1.0.3..
Warning: The package has an extraneous version range for a dependency on an
internal library: htoml-megaparsec -any && ==2.1.0.3, htoml-megaparsec -any &&
==2.1.0.3. This version range includes the current package but isn't needed as
the current package's library will always be used.
Warning:
    This package indirectly depends on multiple versions of the same package. This is very likely to cause a compile failure.
      package htoml-megaparsec (htoml-megaparsec-2.1.0.3-a43bf81dd4ed16baa7df86402245273a070e296a75935859cc970333decde7c9) requires megaparsec-7.0.5-54d1bdf0ee32051801b7c529a69a41a0acc67473704149889aef5337a11842d2
      package htoml-megaparsec (htoml-megaparsec-2.1.0.3) requires megaparsec-7.0.5-9a3f6cbc12071475f205b18cb99b964689c854f3fb71eaf81d216338d2693951
      package megaparsec (megaparsec-7.0.5-54d1bdf0ee32051801b7c529a69a41a0acc67473704149889aef5337a11842d2) requires parser-combinators-1.1.0-4a323801805b1b1e8dc52a5c2286803ad99ec50727aae4428d88e6488abdd583
      package megaparsec (megaparsec-7.0.5-9a3f6cbc12071475f205b18cb99b964689c854f3fb71eaf81d216338d2693951) requires parser-combinators-1.2.0-81eabffcf72fb987500cc5956d9da1d7de5f06d979d7ea0c2614ddbf9c0421ec
Preprocessing library for htoml-megaparsec-2.1.0.3..
Building library for htoml-megaparsec-2.1.0.3..
[1 of 1] Compiling Text.Toml        ( src/Text/Toml.hs, dist/build/Text/Toml.o )

src/Text/Toml.hs:22:16: error:
    â€¢ Couldn't match type â€˜ParseErrorBundle Text e0â€™
                     with â€˜megaparsec-7.0.5:Text.Megaparsec.Error.ParseErrorBundle
                             Text Data.Void.Voidâ€™
```

You can find the full log here:

* https://pastebin.com/w8byLmC1

**To Reproduce**

Cannot reproduce locally. Locally everything works for me.

**Expected behavior**

Successful builds.

**Additional context**

Using `cabal-3.0` on Travis CI.

**Comment by phadej** [comment]
The log says

```
$ cabal --version
cabal-install version 3.0.0.0
```

have you tried with `cabal-install-3.0` locally?

**Comment by chshersh** [comment]
@phadej My mistake. Fixed the `cabal` version in description. Yeah, it works locally with both `cabal-2.4` and `cabal-3.0`.

**Comment by Mikolaj** [comment]
Thank you for reporting. I think this is fixed in cabal 3.4. Please let us know if you can repro this with 3.4 or newer cabals. In any case, closing as a duplicate of #5782.
